openedx/adg/lms/applications/api/tests/test_serializers.py

```diff
@@ -138,3 +142,31 @@ def test_work_experience_serializer_with_valid_data(work_experience_data):  # py
 
     assert serializer.is_valid()
     assert not serializer.errors
+
+
+@pytest.mark.django_db
+def test_reference_serializer_validate_user_application(user_application):  # pylint: disable=redefined-outer-name
+    """
+    Test that the `validate_user_application` method of `ReferenceSerializer` allows adding references only upto the max
+    reference limit against a user application
+    """
+    ReferenceFactory.create_batch(MAX_NUMBER_OF_REFERENCES - 1, user_application=user_application)
+
+    reference_data = {
+        'name': 'Test Name',
+        'position': 'Test Position',
+        'relationship': 'Test Relationship',
+        'phone_number': '112233',
+        'email': 'test@test.com',
+        'user_application': user_application.id
+    }
+
+    serializer = ReferenceSerializer(data=reference_data)
+    assert serializer.is_valid()
+    assert not serializer.errors
+
+    ReferenceFactory(user_application=user_application)
+
+    serializer = ReferenceSerializer(data=reference_data)
+    assert not serializer.is_valid()
+    assert serializer.errors == {'user_application': [MAX_REFERENCE_ERROR_MSG]}
```

Comment:
Let's please save rewriting the last two asserts by parametrizing this test and receive the `is_valid`, `errors` (if applicable) for different cases.

Reply:
These are not two separate test cases, but just one test case. The last two asserts are dependent upon the rest of the code in the test function. The first set of assert statements check the boundary condition, i.e. let us say if max number of references are 2, and there is one reference present, the serializer should validate that reference. Then a new reference is created, so that now there are two references present, i.e. equal to the max limit. The second set of asserts checks that the serializer does not validate a third reference.

If we were parameterize it like you are suggesting, then it would result in unnecessary extra computation. The valid test will create `max - 1` references and then execute the asserts. The invalid test will again create `max-1` references, and another extra reference, and then execute the asserts.

Reply:
It won't really be that much extra computation, and it the recommended way of testing things. We can perhaps take the number of references to be created as a parameter, and then bulk_create that number of references, that would address the extra computation. Secondly, as far as I can tell the first assert and the second assert are not directly dependent on each other i.e either an application will have one reference or two references, it can't be both.

I would still suggest parametrizing this, but I leave the last decision on you.

Reply:
Do you think that setting default value as `undefined` will be a better approach to this according to programatic conventions. According to that, your concern of not assigning null to variables other than those which are meant to hold objects would also be addressed.

Reply:
I've accommodated the suggestion. Thanks for going the extra mile and convince me for it. I believe you are right.